File: mobile/src/main/java/uk/co/appsbystudio/geoshare/login/LoginFragment.java
package uk.co.appsbystudio.geoshare.login;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.marlonmafra.android.widget.EditTextPassword;

import uk.co.appsbystudio.geoshare.MainActivity;
import uk.co.appsbystudio.geoshare.R;

public class LoginFragment extends Fragment{

    private FirebaseAuth firebaseAuth;
    private FirebaseDatabase firebaseDatabase;
    private DatabaseReference databaseReference;

    EditText emailEntry;
    EditTextPassword passwordEntry;
    String email;
    String password;

    public LoginFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);

        firebaseAuth = FirebaseAuth.getInstance();

        emailEntry = (EditText) view.findViewById(R.id.email);
        passwordEntry = (EditTextPassword) view.findViewById(R.id.password);
        Button loginButton = (Button) view.findViewById(R.id.log_in);

        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                validate();
            }
        });

        return view;
    }

    private void validate() {
        email = emailEntry.getText().toString();
        password = passwordEntry.getText().toString();

        boolean cancel = false;
        View focusView = null;

        if (TextUtils.isEmpty(email)) {
            emailEntry.setError(getString(R.string.error_field_required));
            focusView = emailEntry;
            cancel = true;
        }

        if (TextUtils.isEmpty(password)) {
            passwordEntry.setError(getString(R.string.error_field_required));
            focusView = passwordEntry;
            cancel = true;
        }

        if (cancel) {
            focusView.requestFocus();
        } else {
            login();
        }
    }

    private void login() {
        firebaseAuth.signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(getActivity(), new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (!task.isSuccessful()) {
                            if (task.getException() != null) Toast.makeText(getActivity(), task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                        } else {
                            startActivity(new Intent(getActivity(), MainActivity.class));
                            getActivity().finish();
                        }
                    }
                });
    }
}

File: mobile/src/main/java/uk/co/appsbystudio/geoshare/authentication/login/LoginFragment.kt
package uk.co.appsbystudio.geoshare.authentication.login

import android.content.Context
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import kotlinx.android.synthetic.*
import kotlinx.android.synthetic.main.fragment_login.*
import uk.co.appsbystudio.geoshare.R
import uk.co.appsbystudio.geoshare.authentication.AuthActivity
import uk.co.appsbystudio.geoshare.authentication.AuthView
import uk.co.appsbystudio.geoshare.authentication.forgotpassword.ForgotPasswordFragment
import uk.co.appsbystudio.geoshare.authentication.signup.SignupFragment


class LoginFragment : Fragment(), LoginView {

    private var fragmentCallback: AuthView? = null
    private var presenter: LoginPresenter? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_login, container, false)

        presenter = LoginPresenterImpl(this, LoginInteractorImpl())

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        progress_button_login.setOnClickListener {
            presenter?.validate(edit_email_login.text.toString(), edit_password_login.text.toString())
        }

        button_signup_login.setOnClickListener {
            fragmentCallback?.setFragment(SignupFragment())
        }

        button_forgot_password_login.setOnClickListener {
            fragmentCallback?.setFragment(ForgotPasswordFragment())
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)

        try {
            fragmentCallback = context as AuthActivity
        } catch (e: ClassCastException) {
            throw ClassCastException("""${activity.toString()}must implement AuthView""")
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        this.clearFindViewByIdCache()
    }

    override fun setEmailError() {
        edit_email_login.error = resources.getString(R.string.error_field_required)
        edit_email_login.requestFocus()
    }

    override fun setPasswordError() {
        edit_password_login.error = resources.getString(R.string.error_field_required)
        edit_password_login.requestFocus()
    }

    override fun showProgress() {
        progress_button_login.startAnimation()
    }

    override fun hideProgress() {
        progress_button_login.revertAnimation()
    }

    override fun showError(error: String) {
        Toast.makeText(this.context, error, Toast.LENGTH_SHORT).show()
    }

    override fun updateUI() {
        fragmentCallback?.onSuccess()
    }
}
